[Security Solution] Persistent rules table state

maximpn opened the pull request:
**Addresses:** https://github.com/elastic/kibana/issues/140263

## Summary

It implements rules table state persistence. Reopening or reloading of the page loads the same state as well as navigation from another page inside or outside the security plugin.

https://user-images.githubusercontent.com/3775283/201983621-e57e8c64-d5fb-4111-92df-ebe912dada38.mov

## Details

The changes persist the rules table state in the url and the session storage to be able to recover after navigation from another page. The following parameters are persisted

- search term, tags and elastic vs custom filters
- sorting field and sorting direction
- page number **(only in the url)** and page size

Page number is only persisted in the url. Avoiding persistence a page number in the session storage make the UX more predictable that users upon returning to the page expect the first page to be selected. 

Selection state is not persisted since it can lead to URL overflow problem. According to the different resources for example [this one](https://www.geeksforgeeks.org/maximum-length-of-a-url-in-different-browsers/) the maximum length should not exceed 2KB though Chrome supports URLs up to 2MB.

Besides that this PR adds

- unit tests which coved edge cases
- basic Cypress tests

What is not done and will be addressed in the next PRs

- persistence of the selected tab as a path segment in the url
- comprehensive Cypress tests covering edge cases
- getting rid of the back button `< Rules` on the rule details page

## Edge cases

Since parameters in the URL and the session storage can be modified by users it can lead to unexpected results or security risks.
What have been tested and covered so far:

- arbitrary modification of the serialized state which leads to deserialization error (default parameters are used in this case and the state is rewritten with the default parameters)
- filter parameters modification (doesn't cause errors)
- page number modification (isn't not handled in a special way, `EuiBasicTable` display either first of the last page if the value exceeds the limits)
- page size (`EuiBasicTable` loads more records which isn't limited and can cause a performance issue, it **has been addressed** through validation)
- sorting parameters modification (`EuiBasicTable` ignores wrong values)

The observation result is that `EuiBasicTable` gracefully handles the wrong values .

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

maximpn committed:
move rules table filters component next to its siblings

maximpn committed:
preserve rules table filter state in the url

maximpn committed:
add rules table state parameters to the url

maximpn committed:
preserve rules table state to localStorage

maximpn committed:
allow updating initial value

Filter text is initialized in the second render due to reading data in a hook's callback. So having an ability to modify the initial value helps to avoid initializing bugs.

maximpn committed:
split persistence state functionality into initial and sync hooks

maximpn committed:
preserve the active rules tables tab

maximpn committed:
shorten saved state fields

maximpn committed:
add e2e tests

maximpn committed:
remove unused imports

maximpn committed:
fix the case for the cleared storage

maximpn committed:
remove unused import

maximpn committed:
fix the updated import

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

maximpn committed:
validate the saved page size param

maximpn committed:
rename the storage key constant

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_sync_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    updateUrlParam(savedState);
+    storage.set(RULES_TABLE_STATE_STORAGE_KEY, savedState);
```
I have doubts about persisting the current table state in local storage. It provides quite a confusing user experience. For example, if I select table filter in one tab, then open another tab and navigate to the rules table, I'll see the state from the previous tab: 

https://user-images.githubusercontent.com/1938181/202178637-b13492bf-7238-4dc7-9244-2f3b5bce4a23.mov

This behavior is pretty uncommon and could be unexpected from a user standpoint.

On top of that, local storage has a long expiration time. That means if I filtered the table once and then returned to it one month later, I'll see it in the state I left it a month ago, which could be pretty stale. So users will likely have to reset the table filters to their initial state every time they open it. 

My suggestion here is to use either session storage for the table filters or a combination of URL + in-memory state, how it's done in other areas of security solution.

banderror replied:
++ I think using session storage might be a good idea. I'd suggest to test how it works in different browsers because I suspect sessions are not cleared immediately after a user closes their tabs. And I think session storage is a simpler and better persistence mechanism compared to an in-memory state via redux (in our case, of course).

maximpn replied:
I've switched to the session storage and validated the scenarios in Chrome so it keeps a state copy per tab and doesn't restore it after the tab was closed.

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/rules_table_saved_state.ts:
```diff
@@ -0,0 +1,17 @@
+
+export interface RulesTableSavedState {
+  tab?: string;
+  inMemory: boolean;
```
This setting is already stored in local storage; no need to duplicate it. Also, per our discussion during the product meeting, we decided not to make it a part of the URL.

xcrzx reviewed (commented):
I've briefly looked through the implementation and left some initial comments. I think my biggest concern currently is about the usage of local storage. Therefore, I propose discussing alternatives before proceeding with the implementation.

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table_filters/rule_search_field.tsx:
```diff
@@ -0,0 +1,61 @@
+    [setSearchText]
+  );
+
+  useEffect(() => {
+    if (!valueIsDirty) {
+      setSearchText(initialValue);
+    }
+  }, [initialValue, valueIsDirty]);
```
It's bad UX to change the field's value after a user typed something in the field.

maximpn replied:
After the discussion the check fort the field modification by a user was removed.

maximpn committed:
remove memory sorting from the saved state

Memory sorting flag is already preserved in local storage

maximpn commented on x-pack/plugins/security_solution/common/constants.ts:
```diff
@@ -455,6 +455,8 @@ export const NEW_FEATURES_TOUR_STORAGE_KEYS = {
 export const RULE_DETAILS_EXECUTION_LOG_TABLE_SHOW_METRIC_COLUMNS_STORAGE_KEY =
   'securitySolution.ruleDetails.ruleExecutionLog.showMetrics.v8.2';
 
+export const RULES_TABLE_STATE_STORAGE_KEY = 'securitySolution.rulesTable';
```
There are rules table related constants [in that file already](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/constants.ts#L442-L443). It looks quite logical to have `RULES_TABLE_STATE_STORAGE_KEY` here too.

maximpn replied:
After the discussion the rules table related constants used in UI were moved to the related folder.

maximpn committed:
rename rules table url param

Full name url parameter makes it more obvious.

maximpn committed:
preserve table state to session storage instead of local storage

Session storage allows to clear the data after the tab has been closed.

maximpn committed:
avoid syncing default state

maximpn committed:
allow to modify rule search field's initial value

maximpn committed:
move rules tables UI constants to the folder where it's used

maximpn committed:
avoid preserving the active tab

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/rules_table_saved_state.ts:
```diff
@@ -0,0 +1,17 @@
+import type { FilterOptions, SortingOptions } from '../../../../rule_management/logic/types';
+
+export interface RulesTableSavedState {
+  tab?: string;
```
After the discussion the tab persistent logic was removed from this PR. The task related to splitting rules table and rules monitoring table into separate tabs will be addressed in a separate PR.

maximpn committed:
do not provide the active tab to the persistence hooks

jpdjere commented on x-pack/plugins/security_solution/public/plugin.tsx:
```diff
@@ -76,6 +76,7 @@ export class Plugin implements IPlugin<PluginSetup, PluginStart, SetupPlugins, S
   private appUpdater$ = new Subject<AppUpdater>();
 
   private storage = new Storage(localStorage);
+  private sessionStorage = new Storage(sessionStorage);
```
Any specific reason we decided for using `sessionStorage` instead of `localStorage`? I can imagine a user closing a tab, or window, and reopening it, and still want the table to have its state persisted.

maximpn replied:
Right, `I can imagine a user closing a tab, or window, and reopening it, and still want the table to have its state persisted.` this is the scenario which we want to avoid. The local storage persists the state for too long so returning back in a month the user will still have the same state persisted while the filtering and paging data may be outdated and does not show the same rules on the page. @xcrzx commented [here](https://github.com/elastic/kibana/pull/145111#discussion_r1023936654) on this and after the thorough discussion we decided that the session storage fits better.

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.test.ts:
```diff
@@ -0,0 +1,410 @@
+  useRulesTableContext: jest.fn(),
+}));
+
+function mockState({
+  urlState,
+  storageState,
+}: {
+  urlState: RulesTableSavedState | null;
+  storageState: RulesTableSavedState | null;
+}): void {
```
Nit: arrow function

```
const mockState = ({
  urlState,
  storageState,
}: {
  urlState: RulesTableSavedState | null;
  storageState: RulesTableSavedState | null;
}): void => {
```

maximpn replied:
It's rather a matter of taste. I prefer a `function` over an `arrow function` in the global context since it's verbose and easy to read. As a side effect `this` is `undefined` is strict mode.

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.test.ts:
```diff
@@ -0,0 +1,410 @@
+      expect(actions.setPage).toHaveBeenCalledWith(urlSavedState.page);
+      expect(actions.setPerPage).toHaveBeenCalledWith(urlSavedState.perPage);
+    });
+  });
```
Maybe an additional test case to test that in case of partial urlState and storageState, some pieces can be restored from URL and others from storage?

maximpn replied:
Good point, I've added such a test.

jpdjere commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_sync_rules_table_saved_state.test.tsx:
```diff
@@ -0,0 +1,253 @@
+      );
+    });
+  });
+});
```
Same here: at least one case to test that both the URL and storage are updated at the same time when state updated?

maximpn replied:
Good point, I've added such a test.

jpdjere commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/rules_table_persistent_state.cy.ts:
```diff
@@ -0,0 +1,93 @@
+    expectNumberOfRules(1);
+    expectToContainRule('rule 1');
+  });
+
```
Can we add some test cases of maintaining state of other possible filters? `tags`, `page` and `perPage` at least.

maximpn replied:
There is an idea to prepare a test plan for the feature and extend Cypress tests in a separate PR.

maximpn committed:
add a unit tests for recovering the rules table state from the url and the storage

maximpn committed:
add a unit test for syncing the rules table state with the url and the state simultaneously

maximpn committed:
clear the session storage before bulk editing

maximpn commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/bulk_edit_rules.cy.ts:
```diff
@@ -132,6 +132,7 @@ describe('Detection rules, bulk edit', () => {
     login();
   });
   beforeEach(() => {
+    clearSessionStorage();
```
@MadameSheema this looks as a simple and proper fix since a rules table state is persisted in the session storage and interferes to the bulk edit e2e tests.

maximpn committed:
add a comment

maximpn committed:
clear the session storage before rules exporting

xcrzx commented:
It looks like there's an issue in the synchronization between the table state and the URL. The URL query is not being updated when the user unselects the "Elastic rules" filter:

https://user-images.githubusercontent.com/1938181/203122827-ddf635e3-1e3a-43da-928a-10ef1ecf7f95.mov

xcrzx commented:
Similar issue with the search bar:

https://user-images.githubusercontent.com/1938181/203124315-d23e4b4a-a82d-41b3-bd8a-736701a08337.mov

maximpn committed:
make the function to reset the rules table state much more explicit

maximpn committed:
fix syncing of an empty state

maximpn commented:
@xcrzx thank you for the detailed review and testing. The edge case you found was cause by the recent changes regarding syncing only the non default values. I've fixed the problem and modified the unit tests to cover that case.

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_sync_rules_table_saved_state.test.tsx:
```diff
@@ -0,0 +1,296 @@
+      );
+    });
+
+    it('syncs only the sorting order', () => {
+      expectUrlSync(
+        {
+          ...defaultState,
+          sortingOptions: { ...defaultState.sortingOptions, order: 'asc' },
+        },
+        { sort: { order: 'asc' } }
+      );
+    });
```
I find it quite difficult to understand what's being tested here. Could you make these test assertions explicit?

xcrzx reviewed (commented):
Tested it locally, and the implementation works as expected. So previous issues were fixed 👍  The URL/storage synchronization logic also seems correct; I've left only some minor comments regarding tests. But overall, the PR LGTM!

maximpn committed:
simplify e2e tests

maximpn commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/rules_table_persistent_state.cy.ts:
```diff
@@ -0,0 +1,91 @@
+  });
+
+  it('preserved after navigation from the rules details page', () => {
+    visit(DETECTIONS_RULE_MANAGEMENT_URL);
```
It's important to make sure the state isn't persisted in the url so navigation to the "clear" page does the trick. I've moved that step to the `beforeEach` section. `before` section is executed only once per the test.

maximpn commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/rules_table_persistent_state.cy.ts:
```diff
@@ -0,0 +1,91 @@
+  });
+
+  it('preserved after navigation from another page', () => {
+    visit(DETECTIONS_RULE_MANAGEMENT_URL);
```
The same as above, I've moved `visit(SECURITY_DETECTIONS_RULES_URL);` to the `beforeEach` section to make sure the rules table state is cleared in the url.

maximpn commented on x-pack/plugins/security_solution/cypress/e2e/detection_rules/rules_table_persistent_state.cy.ts:
```diff
@@ -0,0 +1,91 @@
+    visit(DETECTIONS_RULE_MANAGEMENT_URL);
+    filterBySearchTerm('rule 1');
+
+    visit(DASHBOARDS_URL);
+    cy.wait(300);
```
`visit(DASHBOARDS_URL);` is important here to accomplish the test. The test itself tests navigation away and having the persistent state when returned back. `cy.wait(300);` just mimics user's behaviour so it looks safe to get rid of it.

maximpn committed:
rename state parameters to improve readability

maximpn committed:
move default rules table constants to a separate file

maximpn committed:
remove unnecessary clear mocks call

maximpn committed:
rename helper functions to improve readability

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.test.ts:
```diff
@@ -0,0 +1,427 @@
+  INITIAL_FILTER_OPTIONS: jest.requireActual('./rules_table_context').INITIAL_FILTER_OPTIONS,
+  INITIAL_SORTING_OPTIONS: jest.requireActual('./rules_table_context').INITIAL_SORTING_OPTIONS,
+  useRulesTableContext: jest.fn(),
+}));
```
I've moved these constants to a separate file. It should be much more logical this way than reexport them.

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.test.ts:
```diff
@@ -0,0 +1,427 @@
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    actions = {
+      setFilterOptions: jest.fn(),
+      setSortingOptions: jest.fn(),
+      setPage: jest.fn(),
+      setPerPage: jest.fn(),
+    };
+
+    (useRulesTableContext as jest.Mock).mockReturnValue({ actions });
```
Unfortunately I need to make expectations on actions and mock actions from `__mock__` folder just don't work. It's clear the mocks are invoked but the tests fail with the error the mocks aren't invoked. The problem lays under the hood, when mocks are exported from another module it seems a new instance is created which breaks tests.

@xcrzx If you know the solution or have another thoughts on this topic it would be great.

xcrzx replied:
Okay, then I think it would be better to mock the entire context with `useRulesTableContextMock.create()`. That will help us keep the mocked object structure in one place, and if something changes, we'll need to update only there. Plus, that mock is strongly typed, which makes it easier to find affected code when types change.

See how a test could be implemented in that case, for example:

https://github.com/elastic/kibana/blob/2efe77b5d847797c6d4763d540a117044fd69a81/x-pack/plugins/security_solution/public/detections/components/rules/rule_switch/index.test.tsx#L92-L106

maximpn committed:
move rules table persisted state mock helper to the mocks folder

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.test.ts:
```diff
@@ -0,0 +1,427 @@
+  storageState,
+}: {
+  urlState: RulesTableSavedState | null;
+  storageState: RulesTableSavedState | null;
```
Done, I've moved and renamed the helper

semd commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_sync_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,69 @@
+      savedState.sort = { ...savedState.sort, order: state.sortingOptions.order };
+    }
+
+    if (state.pagination.page !== DEFAULT_PAGE) {
+      savedState.page = state.pagination.page;
+    }
```
Do we really want to persist the `page`?

I mean, from a UX point of view, after a reload, will users expect/notice the page was persisted, or will they expect it was reset to the first one?

Or, if I navigate to Rules from another page, after the rules table is loaded, does it makes sense to find myself on a page that is not the first one? I don't think it is obvious.

The rest of the parameters make a lot of sense, they are a kind of table "settings", but I don't feel the same way about the `page` parameter, even more considering that the selected rules aren't persisted either.

Just to have a think.

banderror replied:
Great point @semd, we'll think about that 👍

maximpn replied:
@semd you have a valid point though it depends on the usage scenario. One of the problems state persistence solve is returning back from the rule details page. Expectation is the user sees the same page. Another scenario is link sharing where some user opening a link should see the specific page.

Different scenarios like returning from another page inside or outside the security plugin may cause some confusion but it's mitigated by using the session storage.

There is also a suggestion to give users a simple way to clear the table state for example a button next to the filter.

So we need to analyse that scenarios and decide what fits better.

semd replied:
Okay, link sharing could still work with the specific page without persisting the value to the storage. But if you think it makes sense to have it, let's keep it 👍

maximpn replied:
@semd we decided to omit the page number from the session storage after discussion with the team.

machadoum commented:
Hey @maximpn, I checked the code and noticed that you are using `useInitializeUrlParam` slightly different than how it was designed. So, in some scenarios, it behaves differently than in other parts of the app.

**First, let me explain how `useInitializeUrlParam` should be used.**
`useInitializeUrlParam` is designed to be called when the security application loads and not when a page loads. We need it at the application top level because it adds the URL param to the redux store, so every link to rules in the application will contain your URL param.
<img width="1142" alt="Screenshot 2022-11-28 at 10 36 14" src="https://user-images.githubusercontent.com/1490444/204243965-cea82f80-3af2-4a26-93f4-c3ba7d8b94a9.png">

The `onInitialize` callback should persist the data extracted from the URL to the application state. As an example, for the timeline, we inject the data into the timeline store.

By following this pattern, we don't need to use the session store because the status of the application (redux store) is preserved across pages. And when we open any page in another tab, it will carry the URL params.


**About your implementation:**
Instead of persisting the data inside the redux store, you are using session storage, which is slightly less performant but not a problem. You call `useInitializeUrlParam` every time the Rules page loads, which is also somewhat less performant. So, the central issue here is that when users are on the rules page and open the rules link in another tab, it carries all URL params to the new tab. Still, if users are on another page and open the Rules link in another tab, it loses the URL params, which is unexpected.
<img width="1150" alt="Screenshot 2022-11-28 at 11 15 26" src="https://user-images.githubusercontent.com/1490444/204252759-f01a1291-9285-48d1-9490-1829ac842b7c.png">

**Long story short:** You could move `useInitializeUrlParam` to the top of the security solution rendering tree and persist the URL data only once. Like here: https://github.com/elastic/kibana/blob/dbb7f478d32a234a66b877355d9b19df333d4816/x-pack/plugins/security_solution/public/common/hooks/use_url_state.ts#L18

If the current design of `useInitializeUrlParam` and `useUpdateUrlParam` doesn't match your requirement. Another viable option is to read and write to the URL using low-level APIs. We also have created helpers for it: `useGetInitialUrlParamValue` and `useReplaceUrlParams`
https://github.com/elastic/kibana/blob/3c0086b76478f8c011883c0a0ea94354dd0c8f60/x-pack/plugins/security_solution/public/common/utils/global_query_string/helpers.ts#L12

maximpn commented:
Hey @machadoum,

Thank you for the detailed explanation. Basically what you say 

> So, the central issue here is that when users are on the rules page and open the rules link in another tab, it carries all URL params to the new tab. Still, if users are on another page and open the Rules link in another tab, it loses the URL params, which is unexpected.

is how we want the rule table state to persist. A new tab shouldn't carry over the state. Url parameters only make sense on the rules table page. We don't want to display them at the other pages. So it can be considered as filter parameters for a specific page.

On top that we actually have two tables - rules table and rule monitoring table. Both looks very similar and share the same filters but have different columns. So it's only one state instance is persisted but we could need two independent instances in the future.

I'll check if `useGetInitialUrlParamValue` and `useReplaceUrlParams` fit better here.

xcrzx commented:
> `useInitializeUrlParam` is designed to be called when the security application loads and not when a page loads. We need it at the application top level because it adds the URL param to the redux store, so every link to rules in the application will contain your URL param.

> By following this pattern, we don't need to use the session store because the status of the application (redux store) is preserved across pages. And when we open any page in another tab, it will carry the URL params.

Thank you for the explanation, @machadoum! I didn't know that the redux <> URL integration also affects rules management page links. I wonder if all links (including the ones on the rules detail page) automatically pick up URL parameters. If that's the case, we should probably switch to that implementation instead of using the session storage.

maximpn committed:
switch to low level API hooks

maximpn committed:
do not sync page number with the session storage

maximpn committed:
fix a typing issue

maximpn committed:
get rid of the duplicated hook dependencies

machadoum reviewed (approved):
Changes look good to me! 
Thank you for updating the code to use lower-level APIs and for taking the time to improve them! 👏 👏

banderror commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,65 @@
+  useEffect(() => {
+    const { decodedParam: urlState } = getUrlParam();
+
+    const storageState: Partial<RulesTableSavedState> | null = sessionStorage.get(
+      RULES_TABLE_STATE_STORAGE_KEY
+    );
```
Can `sessionStorage.get` throw?

maximpn replied:
It looks safe from the [code](https://github.com/elastic/kibana/blob/main/src/plugins/kibana_utils/public/storage/storage.ts), also `sessionStorage` may only throw [`SecurityError`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage#exceptions) in some really specific use cases which lays out of the scope.

banderror commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,65 @@
+      return;
+    }
+
+    const searchTerm = urlState?.searchTerm ?? storageState?.searchTerm;
+    const showCustomRules = urlState?.showCustomRules ?? storageState?.showCustomRules;
+    const tags = urlState?.tags ?? storageState?.tags;
+    const sorting = urlState?.sort ?? storageState?.sort;
+    const page = urlState?.page ?? storageState?.page;
+    const perPage = urlState?.perPage ?? storageState?.perPage;
```
Nit: If we add validation-on-read, which will make sure that `urlState` and `storageState` actually match the `RulesTableSavedState` interface, this merge could be simplified to

```ts
const restoredState: RulesTableSavedState = { ...storageState, ...urlState };
```

banderror commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/__mocks__/mock_rules_table_persistent_state.ts:
```diff
@@ -0,0 +1,25 @@
+  (useGetInitialUrlParamValue as jest.Mock).mockReturnValue(
+    jest.fn().mockReturnValue({ decodedParam: urlState })
+  );
+  (useKibana as jest.Mock).mockReturnValue({
+    services: { sessionStorage: { get: jest.fn().mockReturnValue(storageState) } },
+  });
```
Not sure about that. This code will completely replace what mocked `useKibana` returns by default which could break a test if it will depend on both session storage and some other service.

I'd suggest to try mocking session storage in `x-pack/plugins/security_solution/public/common/lib/kibana/__mocks__/index.ts` and here it should retrieve the mocked session storage and call `services.sessionStorage.get.mockReturnValue(storageState)`.

maximpn replied:
Correct me if I'm wrong I think you mean spying on items exposed by `useKibana` manual mock which documented [here](https://jestjs.io/docs/es6-class-mocks#spying-on-methods-of-our-class). The problem [it doesn't work](https://stackoverflow.com/questions/62369378/cannot-spy-on-mocked-functions-of-a-class-using-a-manual-mock). If I expose a spy try to mock return value or expect a number of calls it just doesn't catch the necessary calls like a two different instances are created.

banderror reviewed (commented):
This looks great for the basic implementation 👍 

Thanks for cleaning up the code (extracting smaller components like `RuleSearchField`) and the tests, and also adding unit tests for the new hooks. The state persistence looks to be implemented well, but in its basic form, it lacks validation and error handling which I'd like us to add in the next PR.

I left a few minor suggestions. Thank you @maximpn!

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,65 @@
+  } = useKibana();
+
+  useEffect(() => {
+    const { decodedParam: urlState } = getUrlParam();
```
`getUrlParam` shouldn't fail but it depends on the `query-string` external package. It parses a query string if it can or returns an empty object. There are no expected exceptions here. If Rison decoding fails then `decodedParam` is just `null`.

In my opinion it's responsibility of the `useGetInitialUrlParamValue` hook to make sure there are no errors thrown but it looks safe from the code and I'd like to avoid over engineering here.

> What happens when the query string doesn't match the RulesTableSavedState interface?

If it's parsable so it's returned and may not match the TS type. Though the type is optional which means that every field can be missed. This way it's absolutely safe to rename, remove or add new fields. It will lead to just disabling part of the functionality when an outdated link is used.

@banderror if you have some specific suggestions regarding error handling in `useGetInitialUrlParamValue` it'd be nice to hear.

maximpn committed:
implement rules filter as enum

maximpn committed:
split rules table state into url and storage

maximpn committed:
fix wrong typing

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_sync_rules_table_saved_state.test.tsx:
```diff
@@ -0,0 +1,311 @@
+    };
+    const expectedStorageState: RulesTableStorageSavedState = omit(expectedUrlState, 'page');
+
+    (useRulesTableContext as jest.Mock).mockReturnValue({
+      state,
+    });
```
Can the `useRulesTableContextMock` be used here as well? The idea is to make mocks strongly typed. If context types change, we should be able to see all the affected files easily.

Example of how the mock is used in other places:

https://github.com/elastic/kibana/blob/2efe77b5d847797c6d4763d540a117044fd69a81/x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table_utility_bar.test.tsx#L118-L122

maximpn replied:
Right, just pushed it.

maximpn committed:
rename sorting direction to sorting order

maximpn committed:
get rid of `as unknown as`

maximpn committed:
use rules table context from the __mocks__ folder

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,94 @@
+    if (sorting.field || sorting.direction) {
+      actions.setSortingOptions({
+        field: (sorting.field as RulesSortingFields) ?? DEFAULT_SORTING_OPTIONS.field,
+        order: sorting.direction ?? DEFAULT_SORTING_OPTIONS.order,
```
`order` is shorter so I picked it

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/use_initialize_rules_table_saved_state.ts:
```diff
@@ -0,0 +1,94 @@
+
+    if (sorting.field || sorting.direction) {
+      actions.setSortingOptions({
+        field: (sorting.field as RulesSortingFields) ?? DEFAULT_SORTING_OPTIONS.field,
```
It actually highlights another problem that we have two tables (rules and rules monitoring) with slightly different columns but persist only one state. It automatically means not all sorting columns are applicable to the both tables. On top of that we don't already defined enum of supported columns. Providing the wrong sorting column to the table will just disable sorting. Taking it into account I see validation of the sorting field to a restricted set of strings as over-complication. I'd rather implement such validation when/if we split the table saved states.

maximpn committed:
add comments to @ts-expect-errors

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/rules_table/rules_table_saved_state.ts:
```diff
@@ -5,18 +5,43 @@
+export const RulesTableSavedSorting = t.partial({
+  field: t.string,
+  direction: SortOrder,
+});
```
`RulesTableSavedSorting` type is used only in `useInitializeRulesTableSavedState` and `useSyncRulesTableSavedState` to represent the saved state. If somebody changes types declared in the rules table context it doesn't affect the saved sorting options until exposed `setSortingOptions` or `state.sortingOptions` are changed. This way it anyway has to be a mindful change of the saved state.

maximpn committed:
type rules table context state in unit tests

maximpn committed:
handle sessionStorage.get exceptions

maximpn committed:
type sorting field as a union of literals

maximpn committed:
reuse shared types

maximpn committed:
fix typing

xcrzx reviewed (approved):
I looked through the code and tested it locally, fantastic job 🙌 Thank you very much, @maximpn, for improving the rules management table UX and fixing those long-standing and annoying state persistence issues. I believe our users will greatly appreciate this work 😀

banderror reviewed (approved):
The resulting implementation looks great! Thank you for all the fixes @maximpn.
Let's actively ask the rest of the required reviewers to check this PR.

maximpn commented:
@elasticmachine merge upstream

e40pud reviewed (approved):
Alerts area changes LGTM

maximpn commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into persistent-rules-table-state

maximpn commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into persistent-rules-table-state
